### ISSUE TITLE
Implement datastore v1beta3

### ISSUE THREAD

**Issue by dhermes** [GoogleCloudPlatform/gcloud-python#1288]
Current plan of attack:

- [x] Remove `--plugin=protoc-gen-grpc=$(GRPC_PLUGIN)` from "helper" modules that don't use gRPC (e.g. `google.bigtable.v1.bigtable_data`) (this way sub-packages that don't need gRPC can use them) (#1354)
- [x] Ditch shared generated modules for those included with [`protobuf`][6] and [`googleapis-common-protos`][7] (#1353) ~~Move shared (i.e. non-Bigtable specific) modules from `gcloud.bigtable._generated` to `gcloud._generated` (this way all sub-packages can have access to shared protobuf message classes) and then subsequently update the `bigtable` imports to accommodate this change~~
- [x] Pull out [`_get_pb_property_value`][5] in `gcloud.bigtable.cluster` into core (`gcloud._helpers`) and separate out a `_has_field` piece since `pb_message.HasField` doesn't work on non-message fields in `proto3` (#1329)
- [x] Update `Makefile` to reflect changes above (i.e. make sure it can be run without changing the code) (See #1316)
- [x] Update `Makefile` to automatically add the `_foo.proto` files to the repo (See #1317)
- [x] Fold `Batch.add_auto_id_entity` into `Batch.put` (the `auto_id_entity` mutation is removed in favor of just using `insert` with a partial key) (see #1296)
- [x] Loosen `Batch` implementation's reliance on `Mutation` message class (it is used in a `LookupRequest` and their combined structure changes in `v1beta3`) (See #1306, #1319, #1341)
- [x] Remove use of `serializable` in `Client.transaction` and `Transaction` (it is no longer an option in `v1beta3`, so we will curtail it's use in `v1beta2` before the switch) (See #1294)
- [x] Manually partition the `_datastore_v1_pb2.py` module into 3 modules that just import portions of the namespace. These modules (`datastore_pb2.py`, `entity_pb2.py`, `query_pb2.py`) will be according to the membership in the new [proto definitions][4] for `v1beta3`. (See #1297, #1299, #1301, #1328)
- [x] Switch over imports from directly using `_datastore_v1_pb2.py` to using our "shim" imports to mock the structure of `v1beta3`. (See #1297, #1299, #1301, #1328)
- [x] Make sure `API_BASE_URL` for the `Connection` class is used (rather than a parent's version) (See #1293)
- [x] Change `Client.dataset_id` to `Client.project` (in advance of the rename to `PartitionId.project_id` in `v1beta3`) (#1330)
- [x] Change the return type of `Connection.commit` to be a tuple of `index_updates` and `mutation_results` (they were previously on the same result object but are being split apart in `v1beta3`) (See #1314)
- [x] Replace uses of `HasField` for non-message values with `_has_field` (see above) (#1329)
- [x] Remove `v1beta2` generated `pb2` file and old `.proto` definition (can also remove `_datastore_v1_pb2.py` from the `pylintrc_default` ignored file) and delete our "shim" modules (#1355, #1428)
- [x] Update `Makefile` to incorporate protobuf definitions for `v1beta3` (I did this in a [side-project][3], #1353, #1354, #1355, #1428)
- [x] ~~Rewrite our "shim" imports to use the actual `datastore._generated` files for `v1beta3`.~~
- [x] Switch `API_BASE_URL` from `https://www.googleapis.com` to `https://datastore.googleapis.com` and drop `'https://www.googleapis.com/auth/userinfo.email` from the scope list (#1339, #1406)
- [x] Accommodate renames / retyped / removed
    - [x] `CommitRequest.mutation  -->  CommitRequest.mutations` (#1461)
    - [x] `LookupRequest.key  -->  LookupRequest.keys` (~~#1358~~, #1456)
    - [x] `AllocateIdsResponse.key  -->  AllocateIdsResponse.keys` (~~#1358~~, #1456)
    - [x] `AllocateIdsRequest.key  -->  AllocateIdsRequest.keys` (~~#1358~~, #1456)
    - [x] `Key.path_element  -->  Key.path` (~~#1360~~, #1457)
    - [x] `Entity.property  -->  Entity.properties` (#1458)
    - [x] `Query.group_by  -->  Query.distinct_on` (~~#1357~~, #1455)
    - [x] `Query.limit  -->  Query.limit.value` (~~#1357~~, #1455)
    - [x] `QueryResultBatch.entity_result  -->  QueryResultBatch.entity_results` (~~#1357~~, #1455)
    - [x] `PartitionId.namespace  -->  PartitionId.namespace_id` (~~#1359~~, #1452)
    - [x] `PartitionId.dataset_id  -->  PartitionId.project_id` (~~#1359~~, #1452)
    - [x] `Value.indexed  -->  Value.exclude_from_indexes` (~~#1365~~, #1453)
    - [x] `Value.list_value  -->  Value.array_value` (#1460)
    - [x] `Value.null_value` added (#1464)
    - [x] `Value.timestamp_microseconds_value  -->  Value.timestamp_value` (with type change) (#1361, #1459)
    - [x] `Value.geo_point_value` added (#1464)
    - [x] `Value.blob_key_value` removed
    - [x] `CompositeFilter.filter  -->  CompositeFilter.filters` (~~#1356~~, #1454)
    - [x] `CompositeFilter.operation  -->  CompositeFilter.op` (~~#1356~~, #1454)
    - [x] `PropertyFilter.operation  -->  PropertyFilter.op` (~~#1356~~, #1454)
    - [x] `CompositeFilter.AND  -->  CompositeFilter.OPERATOR_UNSPECIFIED` (as default value)  (~~#1356~~, #1454)
    - [x] `ReadOptions.DEFAULT  -->  ReadOptions.READ_CONSISTENCY_UNSPECIFIED` (as default value)  (~~#1356~~, #1454)
- [x] Deal with the fact that `Entity.property` is not a `map` called `Entity.properties` (somewhat different semantics) (#1340)
- [x] Tear out dataset (project) [prefix code][8] (#1466)
- [x] Re-vamp environment variable usage (`dataset_id` no longer needed?) (#1465)
- [x] Remove use of `isolation_level` in `Connection.begin_transaction` (#1343, #1407)

New Features:

- [ ] Add `GqlQuery` support (#304)
- [ ] `QueryResultBatch.skipped_cursor` added (only set when `skipped_results != 0`)
- [ ] `EntityResult.cursor` added (set by the backend when the entity result is part of a `RunQueryResponse.batch.entity_results` response)
- [ ] Parse protobuf [errors][9]

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/94c007dea5a2080cd4dfef0c8d62a6c30d82000e/Makefile#L46
[3]: https://github.com/dhermes/datastore-v1beta3-migration
[4]: https://github.com/google/googleapis/tree/12dd2eafc6ba6d011c4079cbfb246b60f9fdc28d/google/datastore/v1beta3
[5]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/94c007dea5a2080cd4dfef0c8d62a6c30d82000e/gcloud/bigtable/cluster.py#L50-L71
[6]: https://pypi.python.org/pypi/protobuf
[7]: https://pypi.python.org/pypi/googleapis-common-protos
[8]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/3f189536262a7f2046a66c5fb0cb9487ecb0269c/gcloud/datastore/helpers.py#L421
[9]: https://github.com/GoogleCloudPlatform/gcloud-python/pull/1466#issuecomment-183822889

**Comment by pcostell** [comment]
All those project IDs have already been whitelisted for the API.

Just double checking that this is a typo in the issue, but the correct url is:

    https://datastore.googleapis.com/v1beta3/projects/omega-moonlight-697:runQuery

**Comment by dhermes** [comment]
D'oh! Thought I changed it! Thanks for the keen eye

**Comment by dhermes** [comment]
Culprit [found][1], the constructor was relying on the connection base class (non-`datastore` specific)

```python
from gcloud import connection
...
        if api_base_url is None:
            api_base_url = os.getenv(GCD_HOST,
                                     connection.API_BASE_URL)
```

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/883bfe6c83e41ba2a06f18ebca64a933cb6a0eeb/gcloud/datastore/connection.py#L53-L58

**Comment by dhermes** [comment]
For posterity, here is the "retired" original description:

------------

**pcostell** I believe I am ["done"][1] updating for `v1beta3` but am getting 404s when running the system tests with both

```
https://www.googleapis.com/v1beta3/projects/omega-moonlight-697:runQuery
https://www.googleapis.com/v1beta3/projects/s~omega-moonlight-697:runQuery
```

I think it requires a whitelist at this point to use `v1beta3`? I seem to remember a whitelist discussion but can't find any evidence of this.

Project IDs:

```
omega-moonlight-697
precise-truck-742
citric-celerity-697
```

/cc **tseaver **

[1]: https://github.com/dhermes/gcloud-python/tree/datastore-v1beta3

------------

**Comment by dhermes** [comment]
FYI @jgeewax @tseaver I just did a [merge][1] of `master` in the `datastore-v1beta3` branch. I don't know of a better place to surface such comments. Just wanted to keep them in sync / catch any merge conflicts before the happen. Merged perfectly with no issues!

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/commit/36e3bf25a06f61284b666c6ed84483dd6f7eef04

**Comment by dhermes** [comment]
Another FYI, just did another [merge][1]. Only one merge conflict in `system_tests/run_system_test.py
` (due to Bigtable modifying `REQUIREMENTS` and `datastore-v1beta3` just removing it as a variable). Merge conflict easily resolved and merge completed.

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/commit/d8464f02181466bda363025d8f77a6d3085b530d

**Comment by navreetaft** [comment]
Will this update change the user facing API for gcloud-python when interacting with datastore?

Also, will this help with `ndb` being implemented in gcloud-python?  If so, any idea on an ETA for that?  Also, @dhermes commented that 

```ndb saves take about 1/5 the amount of time the gcloud saves do because the RPCs happen directly in GAE instead of via the Cloud Datastore API```

Is this performance gap be addressed by https://github.com/GoogleCloudPlatform/gcloud-python/issues/557?

**Comment by dhermes** [comment]
@navreetaft Essentially yes. The only interface change is `Query.group_by`  to `Query.distinct_on` (to match the changed name in the protobuf). We also add support for geo points and for explicit null values stored and remove support for blob key values. We also plan to add support for GqlQuery.

You can check it out right now at the `datastore-v1beta3` branch. It is 100% working (system tests even pass against the new version).

**Comment by dhermes** [comment]
As for the performance gap, that is in reference to code running within App Engine.

**Comment by navreetaft** [comment]
I see, so the performance gap is due to `ndb` being able to talk directly to DataStore servers, whereas gcloud-python will have to go through some REST gateway (an additional hop)?  Just want to make sure I understand.

**Comment by dhermes** [comment]
Yup. The `ndb` library (inside GAE) talks to Google services via RPCs (remote-procedure calls) while `gcloud-python` (at least when I ran that test in November 2014) was using standard HTTPS requests to just call the datastore API.

**Comment by dhermes** [comment]
@pcostell May have some thoughts / comments here

**Comment by pcostell** [comment]
Yes, right now `ndb` inside GAE will still be faster than `gcloud-python` due to communication via RPC instead of HTTPS requests. `gcloud-python` on `v1beta3` will be faster from GAE because `v1beta3` is more performant than `v1beta2`, however it likely still won't be as fast as `ndb` in GAE for the same reasons above.

**Comment by dhermes** [comment]
Another FYI, just did another [merge][1]. Only merge conflict was the delete of the demos.

/cc @pcostell 

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/commit/f9e4efc3feea3174e203f3aa8b3bb05a0b4d73ee

**Comment by dhermes** [comment]
Another FYI, just did another merge but rebased by mistake, d'oh!

@pcostell Both of my projects got 400's when running the system tests against `v1beta3`

**Comment by pcostell** [comment]
@dhermes My guess this is related to some permissions changes -- now you have to explicitly enable the API (instead of just being part of the whitelist).

You can either login to your app with an @google.com account to do this (look for Cloud Datastore API - NEW in the API manager), or add me as an owner and I can do it for you.

**Comment by pcostell** [comment]
@dhermes Are there any steps that need to happen before we can merge the v1beta3 branch into master and push to pypi?

**Comment by dhermes** [comment]
1. Merge the `datastore-v1beta3` branch into `master`
1. Send a PR to update the version in `setup.py`
1. Push a new tag with the version (Travis pushes to PyPI on tag builds)

**Comment by pcostell** [comment]
@dhermes We are ready for you to merge this in and push to pypi.

**Comment by dhermes** [comment]
w00t.

**Comment by dhermes** [comment]
Did my final [merge][1] of `master` into `datastore-v1beta3`. Only conflict was in `run_system_tests.py` (as predicted).

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/commit/8538095b1283c9ebc23c46d1e69abef15eaea341

**Comment by tseaver** [comment]
@pcostell I can't seem to enable the `v1beta3` API.  I have added you as an owner to my project:  can you figure out what I screwed up?

**Comment by pcostell** [comment]
@tseaver There is an (unfortunately timed) issue with the API service causing this issue. In particular if you disable Datastore v1beta2 (and potentially other APIs), you get left in state where you can't enable APIs. I'll follow up when I get more details.

**Comment by tseaver** [comment]
@pcostell thanks for looking into it.

**Comment by pcostell** [comment]
https://github.com/GoogleCloudPlatform/google-cloud-datastore/issues/97

**Comment by tseaver** [comment]
@pcostell Thank you -- whatever you did for the user in that issue seems to have worked for me, too.
